src/fenics_constitutive/stress_strain.py

```diff
@@ -132,9 +138,198 @@ def strain_from_grad_u(grad_u: np.ndarray, constraint: Constraint) -> np.ndarray
+    to the memory consumption.
+
+    Args:
+        model: 3D model to convert to plane strain.
+
+    Attributes:
+        model: 3D model to convert to plane strain.
+        stress_3d: 3D stress array.
+        tangent_3d: 3D tangent array.
+        grad_del_u_3d: 3D array of gradient of displacement increment.
+    """
+
+    def __init__(self, model: IncrSmallStrainModel) -> None:
+        assert model.constraint == Constraint.FULL
+        self.model = model
+        self.stress_3d = None
+        self.tangent_3d = None
+        self.grad_del_u_3d = None
+
+    @property
+    def constraint(self) -> Constraint:
+        return Constraint.PLANE_STRAIN
+
+    def update(self) -> None:
+        self.model.update()
+
+    def evaluate(
+        self,
+        time: float,
+        del_t: float,
+        grad_del_u: np.ndarray,
+        mandel_stress: np.ndarray,
+        tangent: np.ndarray,
+        history: np.ndarray | dict[str, np.ndarray] | None,
+    ) -> None:
+        n_gauss = int(grad_del_u.size / 4)
+        self.tangent_3d = (
+            np.zeros(6 * 6 * n_gauss) if self.tangent_3d is None else self.tangent_3d
+        )
+        self.stress_3d = (
+            np.zeros(6 * n_gauss) if self.stress_3d is None else self.stress_3d
+        )
+        self.grad_del_u_3d = (
+            np.zeros(9 * n_gauss) if self.grad_del_u_3d is None else self.grad_del_u_3d
+        )
+
+        self._grad_del_u_to_3d(grad_del_u, self.grad_del_u_3d)
+        self._mandel_vector_to_3d(mandel_stress, self.stress_3d)
+
+        self.model.evaluate(
+            time, del_t, self.grad_del_u_3d, self.stress_3d, self.tangent_3d, history
+        )
+        self._tangent_to_2d(self.tangent_3d, tangent)
+        self._mandel_vector_to_2d(self.stress_3d, mandel_stress)
+
+    @property
+    def history_dim(self) -> dict[str, int | tuple[int, int]] | None:
+        return self.model.history_dim
+
+    @df.common.timed("model-conversion-wrapper")
+    def _grad_del_u_to_3d(
+        self, grad_del_u_2d: np.ndarray, grad_del_u_3d: np.ndarray
+    ) -> None:
+        grad_del_u_3d.reshape(-1, 9)[:, 0:2] = grad_del_u_2d.reshape(-1, 4)[:, 0:2]
+        grad_del_u_3d.reshape(-1, 9)[:, 3:5] = grad_del_u_2d.reshape(-1, 4)[:, 2:4]
+
+    @df.common.timed("model-conversion-wrapper")
+    def _mandel_vector_to_3d(
+        self, mandel_vector_2d: np.ndarray, mandel_vector_3d: np.ndarray
+    ) -> np.ndarray:
+        mandel_vector_3d.reshape(-1, 6)[:, 0:4] = mandel_vector_2d.reshape(-1, 4)
+
+    @df.common.timed("model-conversion-wrapper")
+    def _mandel_vector_to_2d(
+        self, mandel_vector_3d: np.ndarray, mandel_vector_2d: np.ndarray
+    ) -> None:
+        mandel_vector_2d.reshape(-1, 4)[:] = mandel_vector_3d.reshape(-1, 6)[:, 0:4]
+
+    @df.common.timed("model-conversion-wrapper")
+    def _tangent_to_2d(
+        self, tangent_3d: np.ndarray, tangent_2d: np.ndarray
+    ) -> np.ndarray:
+        view_2d = tangent_2d.reshape(-1, 16)
```

Comment:
intresting numbers here. But I guess you checked that this are the correct positions ;)

Reply:
I could add a comment here and also determine the indices from an equation to better show how they are constructed

Reply:
If it wouldn't take much time, I appreciate it. But otherwise I trust you :)